inode2prog.cpp

```diff
@@ -73,46 +79,59 @@ int str2int (const char * ptr) {
 	return retval;
 }
 
-// Not sure, but assuming there's no more PID's than go into 64 unsigned bits..
-#define MAX_PID_LENGTH 20
+static std::string read_file (int fd) {
+	char buf[255];
+	std::string content;
 
-char * getprogname (pid_t pid) {
-	int bufsize = 80;
-	char buffer [bufsize];
+	for (int length; (length = read(fd, buf, sizeof(buf))) > 0;) {
+		if (length < 0) {
+			std::fprintf(stderr, "Error reading file: %s\n", std::strerror(errno));
+			std::exit(34);
+		}
+		content.append(buf, length);
+	}
 
-	int maxfilenamelen = 14 + MAX_PID_LENGTH + 1; 
-	char filename[maxfilenamelen];
+	return content;
```

Comment:
OK, so `newnode->name  = getprogname(pid);` below makes copy of the string on heap, and that's freed when `delete current_value` is done on line 134?